clippy_lints/src/missing_fields_in_debug.rs

```diff
@@ -0,0 +1,291 @@
+    /// struct Foo {
+    ///     data: String,
+    ///     // implementation detail
+    ///     hidden_data: i32
+    /// }
+    /// impl fmt::Debug for Foo {
+    ///     fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+    ///         formatter
+    ///             .debug_struct("Foo")
+    ///             .field("data", &self.data)
+    ///             .finish()
+    ///     }
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::fmt;
+    /// struct Foo {
+    ///     data: String,
+    ///     // implementation detail
+    ///     hidden_data: i32
+    /// }
+    /// impl fmt::Debug for Foo {
+    ///     fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+    ///         formatter
+    ///             .debug_struct("Foo")
+    ///             .field("data", &self.data)
+    ///             .finish_non_exhaustive()
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub MISSING_FIELDS_IN_DEBUG,
+    pedantic,
+    "missing fields in manual `Debug` implementation"
+}
+declare_lint_pass!(MissingFieldInDebug => [MISSING_FIELDS_IN_DEBUG]);
+
+const WARN: &str = "manual `Debug` impl does not include all fields";
+const HELP_WILDCARD: &str = "unused field here due to wildcard `_`";
+const HELP_UNUSED_REST: &str = "more unused fields here due to rest pattern `..`";
+const HELP_FIELD_UNUSED: &str = "this field is unused";
+const HELP_FIELD_REFERENCE: &str = "the field referenced by this binding is unused";
+const HELP_INCLUDE_ALL_FIELDS: &str = "consider including all fields in this `Debug` impl";
+const HELP_FINISH_NON_EXHAUSTIVE: &str = "consider calling `.finish_non_exhaustive()` if you intend to ignore fields";
+
+fn report_lints(cx: &LateContext<'_>, span: Span, span_notes: Vec<(Span, &'static str)>) {
+    span_lint_and_then(cx, MISSING_FIELDS_IN_DEBUG, span, WARN, |diag| {
+        for (span, note) in span_notes {
+            diag.span_note(span, note);
+        }
+        diag.help(HELP_INCLUDE_ALL_FIELDS).help(HELP_FINISH_NON_EXHAUSTIVE);
+    });
+}
+
+/// Checks if we should lint in a block of code
+///
+/// The way we check for this condition is by checking if there is
+/// a call to `Formatter::debug_struct` or `Formatter::debug_tuple`
+/// and no call to `.finish_non_exhaustive()`.
+fn should_lint<'tcx>(
+    cx: &LateContext<'tcx>,
+    typeck_results: &TypeckResults<'tcx>,
+    block: impl Visitable<'tcx>,
+) -> bool {
+    let mut has_finish_non_exhaustive = false;
+    let mut has_debug_struct_tuple = false;
+
+    let _: Option<!> = for_each_expr(block, |expr| {
+        let ExprKind::MethodCall(path, recv, ..) = &expr.kind else {
+            return ControlFlow::Continue(());
+        };
+
+        if_chain! {
+            if [sym::debug_struct, sym::debug_tuple].contains(&path.ident.name);
+            if let Some(ty) = typeck_results.expr_ty(recv).peel_refs().ty_adt_def();
+            if match_def_path(cx,ty.did(), &paths::FORMATTER);
```

Comment:
`is_type_diagnostic_item(cx, ty, sym::Formatter)` doesn't work for some reason. It returns false despite `dbg!(ty)` showing that it's `std::fmt::Formatter<'_>`. Not sure why

Reply:
`Formatter` isn't a diagnostic item. You can use `match_type` and add the item to `clippy_utils::paths`.